File: src/main/java/cc/sukazyo/cono/morny/bot/command/Nbnhhsh.java
package cc.sukazyo.cono.morny.bot.command;

import cc.sukazyo.cono.morny.util.tgapi.InputCommand;
import com.pengrad.telegrambot.model.Update;
import com.pengrad.telegrambot.model.request.ParseMode;
import com.pengrad.telegrambot.request.SendMessage;

import cc.sukazyo.cono.morny.MornyCoeur;
import cc.sukazyo.cono.morny.data.NbnhhshQuery;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import static cc.sukazyo.cono.morny.util.CommonConvert.stringsConnecting;
import static cc.sukazyo.cono.morny.util.tgapi.formatting.MsgEscape.escapeHtml;

public class Nbnhhsh implements ITelegramCommand {
	
	@Nonnull @Override public String getName () { return "nbnhhsh"; }
	@Nullable @Override public String[] getAliases () { return null; }
	@Nonnull @Override public String getParamRule () { return "[text]"; }
	@Nonnull @Override public String getDescription () { return "检索文本内 nbnhhsh 词条"; }
	
	@Override
	public void execute (@Nonnull InputCommand command, @Nonnull Update event) {
		
		try {
			
			String queryTarget = "";
			if (event.message().replyToMessage() != null && event.message().replyToMessage().text() != null)
				queryTarget = event.message().replyToMessage().text();
			if (command.hasArgs())
				queryTarget = stringsConnecting(command.getArgs(), " ", 0, command.getArgs().length-1);
			
			NbnhhshQuery.GuessResult response = NbnhhshQuery.sendGuess(queryTarget);
			
			StringBuilder message = new StringBuilder("<a href=\"https://lab.magiconch.com/nbnhhsh/\">## Result of nbnhhsh query :</a>");
			
			for (NbnhhshQuery.Word word : response.words) {
				if (word.trans != null && word.trans.length == 0) word.trans = null;
				if (word.inputting != null && word.inputting.length == 0) word.inputting = null;
				if (word.trans == null && word.inputting == null) continue;
				message.append("\n\n<b>[[ ").append(escapeHtml(word.name)).append(" ]]</b>");
				if (word.trans != null) for (String trans : word.trans) {
					message.append("\n* <i>").append(escapeHtml(trans)).append("</i>");
				}
				if (word.inputting != null) {
					if (word.trans != null) message.append("\n");
					message.append(" maybe:");
					for (String trans : word.inputting) {
						message.append("\n` <i>").append(escapeHtml(trans)).append("</i>");
					}
				}
			}
			
			MornyCoeur.extra().exec(new SendMessage(
					event.message().chat().id(),
					message.toString()
			).parseMode(ParseMode.HTML).replyToMessageId(event.message().messageId()));
			
		} catch (Exception e) {
			MornyCoeur.extra().exec(new SendMessage(
					event.message().chat().id(),
					"[Exception] in query:\n<code>" + escapeHtml(e.getMessage()) + "</code>"
			).parseMode(ParseMode.HTML).replyToMessageId(event.message().messageId()));
		}
		
	}
	
}

File: src/main/scala/cc/sukazyo/cono/morny/bot/command/Nbnhhsh.scala
package cc.sukazyo.cono.morny.bot.command

import cc.sukazyo.cono.morny.MornyCoeur
import cc.sukazyo.cono.morny.data.TelegramStickers
import cc.sukazyo.cono.morny.extra.NbnhhshQuery
import cc.sukazyo.cono.morny.util.tgapi.formatting.TelegramParseEscape.escapeHtml as h
import cc.sukazyo.cono.morny.util.tgapi.InputCommand
import cc.sukazyo.cono.morny.util.tgapi.TelegramExtensions.Bot.exec
import com.pengrad.telegrambot.model.Update
import com.pengrad.telegrambot.model.request.ParseMode
import com.pengrad.telegrambot.request.{SendMessage, SendSticker}
import sttp.client3.{HttpError, SttpClientException}

import java.io.IOException
import scala.language.postfixOps

class Nbnhhsh (using coeur: MornyCoeur) extends ITelegramCommand {
	
	private val NBNHHSH_RESULT_HEAD_HTML =
		// language=html
		"<a href=\"https://lab.magiconch.com/nbnhhsh/\">## Result of nbnhhsh query :</a>"
	
	override val name: String = "nbnhhsh"
	override val aliases: Array[ICommandAlias]|Null = null
	override val paramRule: String = "[text]"
	override val description: String = "检索文本内 nbnhhsh 词条"
	
	override def execute (using command: InputCommand, event: Update): Unit = {
		
		val queryTarget: String =
			if command.args nonEmpty then
				command.args mkString " "
			else if (event.message.replyToMessage != null && event.message.replyToMessage.text != null)
				event.message.replyToMessage.text
			else
				coeur.account exec SendSticker(
					event.message.chat.id,
					TelegramStickers ID_404
				).replyToMessageId(event.message.messageId)
				return;
		
		try {
			
			val queryResp = NbnhhshQuery sendGuess queryTarget
			
			val message = StringBuilder(NBNHHSH_RESULT_HEAD_HTML)
			
			import cc.sukazyo.cono.morny.Log.logger
			logger trace s"**nbnhhsh got len=${queryResp.words.length}"
			for (_word <- queryResp.words) {
				logger trace s"**start for ${_word.name}"
				val _use_trans = (_word.trans ne null) && (_word.trans nonEmpty)
				val _use_inputting = (_word.inputting ne null) && (_word.inputting nonEmpty)
				if (_use_trans || _use_inputting)
					message ++= s"\n\n<b>[[ ${h(_word.name)} ]]</b>"
					logger trace s"**used [${_word.name}]"
					if (_use_trans) for (_trans <- _word.trans)
						message ++= s"\n* <i>${h(_trans)}</i>"
						logger trace s"**used [${_word.name}] used `${_trans}``"
					if (_use_inputting)
						logger trace s"**used [${_word.name}] inputting"
						if (_use_trans)
							message += '\n'
						message ++= " maybe:"
						for (_inputting <- _word.inputting)
							logger trace s"**used [${_word.name}] used-i ${_inputting}"
							message ++= s"\n` <i>${h(_inputting)}</i>"
				logger trace s"**done"
			}
			
			coeur.account exec SendMessage(
				event.message.chat.id,
				message toString
			).parseMode(ParseMode HTML).replyToMessageId(event.message.messageId)
			
		} catch case e: (HttpError[_] | SttpClientException) => {
			coeur.account exec SendMessage(
				event.message.chat.id,
				s"""[Exception] in query:
				   |${h(e.getMessage)}
				   |""".stripMargin
			)
		}
		
	}
	
}
